core/network/rpc_libp2p.hpp

```diff
@@ -0,0 +1,183 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_RPC_LIBP2P_HPP
+#define KAGOME_RPC_LIBP2P_HPP
+
+#include <functional>
+#include <memory>
+
+#include <outcome/outcome.hpp>
+#include "common/buffer.hpp"
+#include "libp2p/basic/readwriter.hpp"
+#include "libp2p/host/host.hpp"
+#include "libp2p/peer/peer_info.hpp"
+#include "libp2p/peer/protocol.hpp"
+
+namespace kagome::network {
+  /**
+   * Works with RPC requests-responses through the Libp2p
+   * @tparam MessageReadWriterT - type of read-writer to be used; for example,
+   * ScaleMessageReadWriter or ProtobufMessageReadWriter
+   */
+  template <typename MessageReadWriterT>
+  struct RPCLibp2p {
+    /**
+     * Read an RPC request and answer with a response
+     * @tparam Request - type of the request to be read
+     * @tparam Response - type of the response to be written
+     * @param read_writer - channel, from which to read and to which to write
+     * @param cb, which is called, when the request is read; it is expected that
+     * this function will return a corresponding response
+     * @param error_cb, which is called, when error happens during read/write or
+     * message processing
+     */
+    template <typename Request, typename Response>
+    static void read(std::shared_ptr<libp2p::basic::ReadWriter> read_writer,
+                     std::function<outcome::result<Response>(Request)> cb,
+                     std::function<void(outcome::result<void>)> error_cb) {
+      auto msg_read_writer =
+          std::make_shared<MessageReadWriterT>(std::move(read_writer));
+      msg_read_writer->template read<Request>(
+          [msg_read_writer, cb = std::move(cb), error_cb = std::move(error_cb)](
+              auto &&request_res) mutable {
+            if (!request_res) {
+              return error_cb(request_res.error());
+            }
+
+            auto response_res = cb(std::move(request_res.value()));
+            if (!response_res) {
+              return error_cb(response_res.error());
+            }
+
+            msg_read_writer->template write<Response>(
+                response_res.value(),
+                [error_cb = std::move(error_cb)](auto &&write_res) {
+                  if (!write_res) {
+                    return error_cb(write_res.error());
+                  }
```

Comment:
same. callback will not be executed :(

Reply:
It was already executed before, now only write error is the think we are interested in